### ISSUE TITLE
react/no-typos: Cannot read property 'name' of undefined

### ISSUE THREAD

**Issue by lynxtaa** [yannickcr/eslint-plugin-react#1373]
These two lines cause a crash with a rule *no-typos* enabled:
```javascript
const { a } = { ...{ a: { z: 1 } } }
console.log(a.z)
```
I'm using eslint@4.4.1 and eslint-plugin-react@7.2.1.

```
Cannot read property 'name' of undefined
TypeError: Cannot read property 'name' of undefined
    at Object.getRelatedComponent (/mnt/d/react-rubin/node_modules/eslint-plugin-react/lib/util/Components.js:531:46)
    at Linter.MemberExpression (/mnt/d/react-rubin/node_modules/eslint-plugin-react/lib/rules/no-typos.js:69:40)
    at emitOne (events.js:120:20)
    at Linter.emit (events.js:210:7)
    at NodeEventGenerator.applySelector (/mnt/d/react-rubin/node_modules/eslint/lib/util/node-event-generator.js:265:26)
    at NodeEventGenerator.applySelectors (/mnt/d/react-rubin/node_modules/eslint/lib/util/node-event-generator.js:294:22)
    at NodeEventGenerator.enterNode (/mnt/d/react-rubin/node_modules/eslint/lib/util/node-event-generator.js:308:14)
    at CodePathAnalyzer.enterNode (/mnt/d/react-rubin/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:602:23)
    at Traverser.enter (/mnt/d/react-rubin/node_modules/eslint/lib/linter.js:925:36)
    at Traverser.__execute (/mnt/d/react-rubin/node_modules/estraverse/estraverse.js:397:31)
```

**Comment by LostCrew** [comment]
I'm currently experiencing the same issue. Let me know if you need more details on my setup.

**Comment by yannickcr** [comment]
Fixed in #1372, should be in next patch release.

**Comment by jseminck** [comment]
@LostCrew actually it could be good if you could find out the code that breaks. Just want to make sure the current fix covers your case too.

**Comment by cheton** [comment]
I got this error with following setup:

* eslint@4.5.0
* eslint-plugin-jsx-a11y@5.1.1
* eslint-plugin-react@7.3.0

```
Cannot read property 'name' of undefined
TypeError: Cannot read property 'name' of undefined
    at Object.utils.isExplicitComponent.getRelatedComponent (/Users/cheton/github/cncjs/cncjs/node_modules/eslint-plugin-react/lib/util/Components.js:540:46)
    at Linter.node.name.PROP_TYPES.some.args.elements.forEach.node.properties.forEach.STATIC_CLASS_PROPERTIES.forEach.LIFECYCLE_METHODS.forEach.MemberExpression (/Users/cheton/github/cncjs/cncjs/node_modules/eslint-plugin-react/lib/rules/no-typos.js:120:40)
    at emitOne (events.js:82:20)
    at Linter.emit (events.js:169:7)
    at NodeEventGenerator.applySelector (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/lib/util/node-event-generator.js:265:26)
    at NodeEventGenerator.applySelectors (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/lib/util/node-event-generator.js:294:22)
    at NodeEventGenerator.enterNode (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/lib/util/node-event-generator.js:308:14)
    at CodePathAnalyzer.enterNode (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:602:23)
    at Traverser.enter (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/lib/linter.js:925:36)
    at Traverser.__execute (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/node_modules/estraverse/estraverse.js:397:31)
    at Traverser.traverse (/Users/cheton/github/cncjs/cncjs/node_modules/eslint/node_modules/estraverse/estraverse.js:501:28)
```

**Comment by jseminck** [comment]
@cheton any chance you can provide an example of the code that breaks it? I understand it may not be easy to find, but it would be very useful!

**Comment by cheton** [comment]
After narrowing down over 500+ files, I discovered a small file that could cause the error.

https://github.com/cncjs/cncjs/blob/master/src/app/api/api.version.js
```js
import url from 'url';
import registryUrl from 'registry-url';
import registryAuthToken from 'registry-auth-token';
import request from 'superagent';
import pkg from '../../package.json';
import {
    ERR_INTERNAL_SERVER_ERROR
} from '../constants';

export const getLatestVersion = (req, res) => {
    const scope = pkg.name.split('/')[0];
    const regUrl = registryUrl(scope);
    const pkgUrl = url.resolve(regUrl, encodeURIComponent(pkg.name).replace(/^%40/, '@'));
    const authInfo = registryAuthToken(regUrl);
    const headers = {};

    if (authInfo) {
        headers.Authorization = `${authInfo.type} ${authInfo.token}`;
    }

    request
        .get(pkgUrl)
        .set(headers)
        .end((err, _res) => {
            if (err) {
                res.status(ERR_INTERNAL_SERVER_ERROR).send({
                    msg: `Failed to connect to ${pkgUrl}: code=${err.code}`
                });
                return;
            }

            const { body: data = {} } = { ..._res };
            data.time = data.time || {};
            data['dist-tags'] = data['dist-tags'] || {};
            data.versions = data.versions || {};

            const time = data.time[latest];
            const latest = data['dist-tags'].latest;
            const {
                name,
                version,
                description,
                homepage
            } = { ...data.versions[latest] };

            res.send({ time, name, version, description, homepage });
        });
};
```

**Comment by cheton** [comment]
This will cause a crash:
```js
const f = (err, res) => {
    const { body: data = {} } = { ...res };
    data.time = data.time || {};
    data['dist-tags'] = data['dist-tags'] || {};
    data.versions = data.versions || {};
};
```

**Comment by jseminck** [comment]
Thanks a lot. I'll try to look into this asap.

In short I think we should start to try to find a way to limit the `no-typos` rule to files that actually contain React.Components. Currently it's visiting **al MemberExpressions** which is perhaps a bit overkill.

**Comment by jseminck** [comment]
@cheton PR is up. If you can, please have a look. Thanks a lot for reporting the issue and taking the time to find a snippet to reproduce the issue. 🎉

**Comment by jseminck** [comment]
@cheton have you had a change to try 7.4.0-rc0? I haven't found any issues anymore. But would be great to confirm if your issue is fixed as well.

**Comment by cheton** [comment]
I can confirm that it's fixed in 7.4.0-rc.0. Thank you!
